packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -395,6 +395,111 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
+      const dataWithoutMissingRelationsPromise = Object.entries(sanitizedSchemaAttributes).reduce(
+        async (
+          previousRelationAttributesPromise: Promise<Record<string, unknown>>,
+          [name, attribute]: [string, Schema.Attribute.AnyAttribute]
+        ) => {
+          const previousRelationAttributes = await previousRelationAttributesPromise;
+
+          const relationData = version.data[name];
+          if (relationData === null) {
+            return previousRelationAttributes;
+          }
+
+          if (
+            attribute.type === 'relation' &&
+            // TODO: handle polymorphic relations
+            attribute.relation !== 'morphToOne' &&
+            attribute.relation !== 'morphToMany'
+          ) {
+            if (Array.isArray(relationData)) {
+              if (relationData.length === 0) return previousRelationAttributes;
+
+              const existingAndMissingRelations = await Promise.all(
+                relationData.map((relation) => {
+                  return strapi.documents(attribute.target).findOne({
+                    documentId: relation.documentId,
+                    locale: relation.locale || undefined,
+                  });
+                })
+              );
+              const existingRelations = existingAndMissingRelations.filter(
+                (relation) => relation !== null
+              ) as Modules.Documents.AnyDocument[];
+
+              previousRelationAttributes[name] = existingRelations;
+            } else {
+              const existingRelation = await strapi.documents(attribute.target).findOne({
+                documentId: relationData.documentId,
+                locale: relationData.locale || undefined,
+              });
+
+              if (!existingRelation) {
+                previousRelationAttributes[name] = null;
+              }
+            }
+          }
+
+          if (attribute.type === 'media') {
+            if (attribute.multiple) {
+              const existingAndMissingMedias = await Promise.all(
+                relationData.map((media) => {
+                  return strapi.db
+                    .query('plugin::upload.file')
+                    .findOne({ where: { id: media.id } });
+                })
+              );
+
+              const existingMedias = existingAndMissingMedias.filter((media) => media != null);
+              previousRelationAttributes[name] = existingMedias;
+            } else {
+              const existingMedia = await strapi.db
+                .query('plugin::upload.file')
+                .findOne({ where: { id: version.data[name].id } });
+
+              if (!existingMedia) {
+                previousRelationAttributes[name] = null;
+              }
+            }
+          }
+
+          return previousRelationAttributes;
+        },
+        // Clone to avoid mutating the original version data
+        Promise.resolve(structuredClone(dataWithoutAddedAttributes))
```

Comment:
ℹ️ `structuredClone` is a deep clone and supported since node 17. Strapi supports node >= 18. A spread may actually be sufficient though. lmk what y'all think.